### PR TITLE
[App] Format client error ApiExceptions without a traceback

### PR DESCRIPTION
**Pull request by luca3rd**
## What does this PR do?

The lightning CLI uses the lightning APIs to create, read, update, and delete resources such as apps and clusters.

If an API call fails due to client error (e.g. invalid input, unauthorized, unauthenticated) the API will return an HTTP status code in the 400s, along with a custom message in the response body.

The CLI (usually) does not directly handle these exceptions. As a result the exception bubbles up through the click framework (which does not recognize the exception type) and to the python runtime, which produces a long traceback and dumps the raw exception to the user.

This PR fixes the user experience. When our API fails on a client error, the CLI will display the message from the server without a traceback.

#### Before
```
➜ lightning create cluster \
        lf-001 \
        --external-id blablabla \
        --role-arn blablabla \
        --edit-before-creation
Traceback (most recent call last):
  File "/Users/lucafurst/.pyenv/versions/lightning-2/bin/lightning", line 33, in <module>
    sys.exit(load_entry_point('lightning', 'console_scripts', 'lightning')())
  File "/Users/lucafurst/go/src/github.com/Lightning-AI/lightning/src/lightning_app/cli/lightning_cli.py", line 79, in main
    _main()
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/lucafurst/go/src/github.com/Lightning-AI/lightning/src/lightning_app/cli/lightning_cli_create.py", line 79, in create_cluster
    cluster_manager.create(
  File "/Users/lucafurst/go/src/github.com/Lightning-AI/lightning/src/lightning_app/cli/cmd_clusters.py", line 136, in create
    resp = self.api_client.cluster_service_create_cluster(body=new_body)
  File "/Users/lucafurst/go/src/github.com/Lightning-AI/lightning/src/lightning_app/utilities/network.py", line 92, in wrapped
    raise e
  File "/Users/lucafurst/go/src/github.com/Lightning-AI/lightning/src/lightning_app/utilities/network.py", line 80, in wrapped
    return func(*args, **kwargs)
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/lightning_cloud/openapi/api/cluster_service_api.py", line 66, in cluster_service_create_cluster
    (data) = self.cluster_service_create_cluster_with_http_info(
  File "/Users/lucafurst/go/src/github.com/Lightning-AI/lightning/src/lightning_app/utilities/network.py", line 92, in wrapped
    raise e
  File "/Users/lucafurst/go/src/github.com/Lightning-AI/lightning/src/lightning_app/utilities/network.py", line 80, in wrapped
    return func(*args, **kwargs)
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/lightning_cloud/openapi/api/cluster_service_api.py", line 133, in cluster_service_create_cluster_with_http_info
    return self.api_client.call_api(
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/lightning_cloud/openapi/api_client.py", line 362, in call_api
    return self.__call_api(resource_path, method, path_params,
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/lightning_cloud/openapi/api_client.py", line 173, in __call_api
    response_data = self.request(method,
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/lightning_cloud/rest_client.py", line 75, in wrap
    raise err
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/lightning_cloud/rest_client.py", line 69, in wrap
    response = func(*args, **kwargs)
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/lightning_cloud/openapi/api_client.py", line 408, in request
    return self.rest_client.POST(url,
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/lightning_cloud/openapi/rest.py", line 307, in POST
    return self.request("POST",
  File "/Users/lucafurst/.pyenv/versions/3.8.12/envs/lightning-2/lib/python3.8/site-packages/lightning_cloud/openapi/rest.py", line 237, in request
    raise ApiException(http_resp=r)
lightning_cloud.openapi.rest.ApiException: (400)
Reason: Bad Request
HTTP response headers: HTTPHeaderDict({'access-control-allow-headers': 'Content-Type,X-Grid-Key,X-Grid-User,Authorization', 'access-control-allow-methods': 'GET,POST,PUT,PATCH,DELETE,HEAD,OPTIONS', 'access-control-allow-origin': '*', 'content-type': 'application/json', 'vary': 'Accept-Encoding', 'date': 'Tue, 25 Oct 2022 23:37:00 GMT', 'content-length': '54', 'x-envoy-upstream-service-time': '2', 'server': 'envoy'})
HTTP response body: b'{"code":3, "message":"name is required", "details":[]}'
```

#### After
```bash
➜ lightning create cluster \
        lf-001 \
        --external-id blablabla \
        --role-arn blablabla \
        --edit-before-creation
Error: name is required
```

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/utilities/exception_handlers.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+from json import JSONDecodeError, loads
```
Comment (awaelchli) [inline src/lightning_app/utilities/exception_handlers.py:1]: How about we move it to utilities/exceptions.py?

**Review by awaelchli — APPROVED** [review]
Great improvement!

File: src/lightning_app/utilities/exception_handlers.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+from typing import Any
+
+from click import ClickException, Context, Group
+from lightning_cloud.openapi.rest import ApiException
+
+
+class ApiExceptionHandler(Group):
```
Comment (awaelchli) [inline src/lightning_app/utilities/exception_handlers.py:8]: Let's make this class protected? We don't expect users to subclass it.

**Review by tchaton — APPROVED** [review]
Super neat !

**Comment by nmiculinic** [comment]
Love it!

**Pull request merged**